src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/DynamicInterfaceCastableImplementation.cs

```diff
@@ -0,0 +1,127 @@
+            DynamicInterfaceCastableImplementationUnsupported,
+            InterfaceMembersMissingImplementation,
+            MembersDeclaredOnImplementationTypeMustBeSealed);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
+            context.RegisterCompilationStartAction(context =>
+            {
+                if (context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemRuntimeInteropServicesDynamicInterfaceCastableImplementationAttribute, out INamedTypeSymbol? dynamicInterfaceCastableImplementationAttribute))
+                {
+                    context.RegisterSymbolAction(context => AnalyzeType(context, dynamicInterfaceCastableImplementationAttribute), SymbolKind.NamedType);
+                }
+            });
+        }
+
+        private static void AnalyzeType(SymbolAnalysisContext context, INamedTypeSymbol dynamicInterfaceCastableImplementationAttribute)
+        {
+            INamedTypeSymbol targetType = (INamedTypeSymbol)context.Symbol;
+
+            if (targetType.TypeKind != TypeKind.Interface)
+            {
+                return;
+            }
+
+            if (!targetType.HasAttribute(dynamicInterfaceCastableImplementationAttribute))
+            {
+                return;
+            }
+
+            // Default Interface Methods are required to provide an IDynamicInterfaceCastable implementation type.
+            // Since Visual Basic does not support DIMs, an implementation type cannot be correctly provided in VB.
+            if (context.Compilation.Language == LanguageNames.VisualBasic)
+            {
+                context.ReportDiagnostic(targetType.CreateDiagnostic(DynamicInterfaceCastableImplementationUnsupported));
+                return;
+            }
+
+            bool missingMethodImplementations = false;
+            foreach (var iface in targetType.AllInterfaces)
+            {
+                foreach (var member in iface.GetMembers())
+                {
+                    if (!member.IsStatic
+                        && context.Compilation.IsSymbolAccessibleWithin(member, targetType)
+                        && targetType.FindImplementationForInterfaceMember(member) is null)
+                    {
+                        missingMethodImplementations = true;
+                        break;
+                    }
+                }
+            }
+
+            if (missingMethodImplementations)
+            {
+                context.ReportDiagnostic(targetType.CreateDiagnostic(InterfaceMembersMissingImplementation, targetType.ToDisplayString()));
+            }
+
+            foreach (var member in targetType.GetMembers())
+            {
+                if (member.IsVirtual || member.IsAbstract)
+                {
```

Comment:
💭 Would this be more readable if it focused on `sealed`?

```suggestion
            {
                if (!member.IsSealed)
                {
```

Reply:
We can't use `IsSealed` here. `IsSealed` is based on the rules for emitting the `final` bit for the member in metadata, but `sealed` DIM members don't set that bit, they just don't set the `virtual` or `abstract` bits.

Reply:
SharpLab link for reference: https://sharplab.io/#v2:C4LglgNgNAJiDUAfAAgJgIwFgBQyDMABGAHbACmATgGYCGAxmQQJIEDeOBnB+BAzmTQhkY3ACwEAsgAoAlGw5cAvjmW5CaAgCF52Lt0IA3MBWABXQWMmyde1ap4aAwgRBabXHgHsDlCmBiM/ILCltJy7LpKKkA==